packages/kolibri-tools/lib/i18n/auditMessages.js

```diff
@@ -0,0 +1,105 @@
+const fs = require('fs');
+const path = require('path');
+const mkdirp = require('mkdirp');
+const { parse } = require('csv-parse/sync');
+const createCsvWriter = require('csv-writer').createObjectCsvWriter;
+const logging = require('../logging');
+const { getAllMessagesFromFilePath } = require('./astUtils');
+const { checkForDuplicateIds, forEachPathInfo } = require('./utils');
+
+// Instantiates the CSV data and writes to a file.
+function writeAuditToCSV(audit, outputFile) {
+  // Ensure we have a {localePath}/audit directory available.
+  mkdirp.sync(path.dirname(outputFile));
+
+  const csvWriter = createCsvWriter({
+    path: outputFile,
+    header: [
+      { id: 'message', title: 'Message' },
+      { id: 'missing', title: 'Missing?' },
+      { id: 'duplicate', title: 'Duplicated?' },
+      { id: 'messageIds', title: 'Message Ids' },
+      { id: 'namespaces', title: 'Namespaces' },
+    ],
+  });
+  csvWriter
+    .writeRecords(
+      audit.map(a => ({
+        ...a,
+        messageIds: a.messageIds.join(', '),
+        namespaces: a.namespaces.join(', '),
+      }))
+    )
+    .then(() => logging.log(`Audit CSV written to ${outputFile}`));
+}
+
+module.exports = function(pathInfo, ignore, dittoFilePaths, outputFile, verbose) {
+  // An object for storing our messages.
+  const extractedMessages = {};
+  forEachPathInfo(pathInfo, pathData => {
+    const namespace = pathData.namespace;
+    if (!extractedMessages[namespace]) {
+      const filePathMessages = getAllMessagesFromFilePath(pathData.moduleFilePath, ignore, verbose);
+      for (const otherNamespace in extractedMessages) {
+        const nameSpaceMessages = extractedMessages[otherNamespace];
+        if (checkForDuplicateIds(nameSpaceMessages, filePathMessages)) {
+          logging.error(
+            `Duplicate message ids across namespaces ${namespace} and ${otherNamespace}`
+          );
```

Comment:
> Checks for duplicate message ids whenever we are about to merge two message objects into each other

Is this primarily for us to find where we can consider making `common` a particular message because they share message ids?

Reply:
No, this is indicative of an error - because we ultimately merge all the messageIds into a single lookup when we pull in from crowdin, so they should be unique across all namespaces.